File: classy/src/make_fake.py
"""
Make a list of fake sources given a reference image and the image to plant into.
"""
import os
import sys

import numpy
import orbit_utils
from astropy import wcs
from astropy.coordinates import SkyCoord
from astropy.io import fits
from astropy.table import Table
from astropy.time import Time


def M_at_jday(M, epoch_M, a, jday) -> float:
    """
    Given the current mean anomaly M at epoch_M and the semi-major axis, compute the mean anomaly at jday
    :param M:
    :param epoch_M:
    :param a:
    :param jday:
    :return:
    """

    nu = 2 * numpy.pi / 365.25
    return (M + (jday - epoch_M) * (nu / a ** 1.5)) % (2 * numpy.pi)


ELEMENT_RANGES = {
    'a': (30, 100),
    'e': (0, 0.5),
    'inc': (0, numpy.pi / 2.0),
    'node': (0, 2 * numpy.pi),
    'peri': (0, 2 * numpy.pi),
    'M': (0, 2 * numpy.pi) }


def build_plant_list_catalog(image_filename: str,
                             reference_epoch: int = 2459792,
                             element_ranges: dict = None,
                             mag_limit: float = 27.5,
                             gb: float = 0.15,
                             ext: str = ".plantList") -> None:
    """
    Given the reference image compute the locations of artificial sourcess on the destination image.
    :param image_filename: filename of FITS image to generate a plantlist for.
    :param reference_epoch: epoch (JD) of the reference date for 'M' and used to set the seed of random generator for elements.
    :param element_ranges: ranges of a/e/i/Omega/omega/M to draw from
    :param mag_limit: limiting magnitude to plant source to
    :param gb: slope of phase function in Bowel formulasim that converts H to mag
    :param ext: extension of filename for plantList output, nominall Image_filename.plantList

    The default reference epoch (integer Julian Day) is 2459792 (2022-07-31T12:00:00.000) , this is used to seed the random number
    generator that produces the catalog that will include sources on the current image.   The value of 'M' generated is take to be
    at this epoch and M is propogated to the JD of image_filename
    :return:
    """

    if element_ranges is None:
        element_ranges = ELEMENT_RANGES

    orbits = {}
    for element in element_ranges.keys():
        orbits[element] = []
    orbits['H']  = []

    max_iters = 1E8
    max_objects = 10000
    obs_code = 500

    wcs_dict = {}
    jd_dict = {}
    print(image_filename)
    with fits.open(image_filename) as hdulist:
        extno = 0
        if hdulist[extno].header['NAXIS'] == 0:
            extno = 1
        wcs_dict[image_filename] = wcs.WCS(hdulist[extno].header)
        jd_dict[image_filename] = Time(hdulist[extno].header['MJD-OBS'], format='mjd').jd

    # initialize the random number generator using the JD of the reference frame.
    rnd_gen = numpy.random.default_rng(reference_epoch)

    # compute the location of the observer on the date image_filename was taken.
    obs_pos, vel, ros, ierr = orbit_utils.obspos(obs_code, jd_dict[image_filename])

    nobjects = 0
    niters = 0
    while nobjects < max_objects and niters < max_iters:
        niters += 1
        orbit = {}
        for element in element_ranges:
            orbit[element] = rnd_gen.uniform(element_ranges[element][0],
                                             element_ranges[element][1],
                                             1)
        # Compute the location of the sources on the image_filename date.
        orbit['M'] = M_at_jday(orbit['M'], reference_epoch, orbit['a'], jd_dict[image_filename])
        pos = orbit_utils.pos_cart(orbit['a'],
                                   orbit['e'],
                                   orbit['inc'],
                                   orbit['node'],
                                   orbit['peri'],
                                   orbit['M'])
        delta, RA, DEC, ierr = orbit_utils.radececlxv(pos, obs_pos)
        if ierr != 0 :
            raise OSError(f'Failed to get RA/DEC for {orbit}')
        coord = SkyCoord(RA, DEC, distance=delta, units=('radian', 'radian', 'au'))
        # check if this sources is within the image we are makeing a plantList for.
        if not wcs_dict[image_filename].footprint_contains(coord):
            continue
        nobjects += 1
        r = numpy.sqrt(pos[0] ** 2, pos[1] ** 2 + pos[2] ** 2)
        h_max = mag_limit - 10 * numpy.log10(r)
        h = rnd_gen.uniform(5, h_max)
        orbit['h'] = h
        alpha, mag, ierr = orbit_utils.appmag(r, delta, ros, h, gb)
        if ierr != 0:
            raise OSError(f"Computing AppMag failed for {orbit}")
        orbit['alpha'] =  alpha
        orbit['mag'] = mag

        for element in orbits.keys():
            orbits[element].append(orbit[element])

    catalog_filename = f"{os.path.splitext(image_filename)[0]}.{ext}"
    Table(orbits).write(catalog_filename)


def main():
    dest_image = sys.argv[1]
    print(dest_image)
    build_plant_list_catalog(dest_image)


if __name__ == '__main__':
    main()

File: fortran_code/MakeFake.f
C-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*
      include 'ran3.f'
      
      program XX

      implicit none

      integer(kind=4) obs_code, ierr, niter, nobjects, object_number
      integer(kind=4) output_lun, max_objects, max_iters
      integer(kind=4) field_number, i, j, version, k, seed


      real(kind=8) Pi, TwoPi, drad, mu, theta, phi, gi
      real(kind=8) gb, alpha, MA, ran3
      real(kind=8) a_limits(2), e_limits(2), i_limits(2)
      real(kind=8) h_limits(2), m_limits(2)
      real(kind=8) group


      parameter (Pi = 3.141592653589793238_8, TwoPi = 2.0d0*Pi,
     $     drad = Pi/180.0d0, mu = TwoPi**2)
      parameter (gb = 0.15, version=5)
      
      real(kind=8) obs_jday, fake_jday
      real(kind=8) a, e, inc, node, peri, M, H, pos(6), obs_pos(3)
      real(kind=8) vel(3)
      real(kind=8) epoch_M
      real(kind=8) ros, fake_obs_pos(3), fake_ros, dec_cen
      real(kind=8) ra, dec, delta, radius, mag, ddec, dra, ra_cen
      real(kind=8) fake_ra, fake_dec, fake_delta, fake_mag, fake_M
      real(kind=8) fake_ra2, fake_dec2, fake_delta2, fake_mag2, fake_M2
      real(kind=8) fake_obs_pos2(3), fake_ros2
      real(kind=8) obs_ra, obs_dec, obs_radius, cos_obs_dec
      real(kind=8) r, ran_mag
      character in_str*120
      character output_filename*120, field*120
      character*80 message

      max_iters = 1E9
      obs_code = 500

      CALL parse_args(field, ra_cen, dec_cen, obs_jday, 
     $                fake_jday, max_objects)

C     field == name of field to make fakes for, can be any string
C     ra_cen == central RA of region to get fake sources for (deg)
C     dec_cen == central DEC of region to get fake sources for (deg)
C     obs_jday == julian date that is the opposition date for this field centre
C     fkae_jday == date to generate source positions for
C     max_objects = number of objects to generate for this field.
      
      epoch_M = obs_jday

C     set the random seed to the same value for all runs, this is a
C     realization of the Kuiper belt
      seed=-123456789
      r=ran3(seed)


C     compute the cos_of_dec once as we need this many times.
      cos_obs_dec = cos(dec_cen*drad)

C     For each field we pull all the sources that are within 1.2
C     degrees. This ensures we can plant sources in every ccd of mosaic.
      obs_radius = 1.2*drad

C     create the output file name string, little finiky. 
      do i = 1, len(in_str)
         in_str(i:i) = ' '
         output_filename(i:i) = ' '
      end do

      write(in_str, 7002) "PlantList",field,
     $        ra_cen,dec_cen,fake_jday,".txt"
 7002 format (2(a10,"_"),3(f15.4,"_"),a4)

      j = 1
      do i = 1, len(in_str) 
         if (in_str(i:i) .ne. ' ') then
            output_filename(j:j) = in_str(i:i)
            j = j+1
         endif
      end do
      output_filename = trim(output_filename)
      write(6,*) "Writing to "//output_filename

      output_lun = 11
      open (unit=output_lun, file=output_filename, status='new')

C     Get observatory location for primary location
      call ObsPos (obs_code, obs_jday, obs_pos, 
     $     vel, ros, ierr)

C     Get the observer location at this fake date.
      call ObsPos (obs_code, fake_jday, fake_obs_pos,
     $     vel, fake_ros, ierr)

C     Get the observer location at fake date + 1 day (for rates)
      call ObsPos (obs_code, fake_jday+1, fake_obs_pos2,
     $     vel, fake_ros2, ierr)

      write(output_lun, 9002) ClassyMakeFakeV5.f'obj_id', 'RA', 'DEC', 'Delta',
     $     'mag', 'a', 
     $     'e', 'inc',
     $     'node', 'peri', 'M', 'H', 'dra_arc',
     $     'ddec_arc', 'r_sun', 'gi'

C     Loop over making fake objects until we have max_objects in frame
      niter = 1
      nobjects = 1
      do while ((nobjects .le. max_objects) .and. 
     $     (niter .lt. max_iters))
         
         a_limits = (/30.0_8,220.0_8/)
         e_limits = (/0.0_8,0.999_8/)
         i_limits = (/0.0_8,Pi/2.0/)
         m_limits = (/22.0_8,28.0_8/)
         group = ran3(seed)
         if (group .lt. 0.8 .and. group .gt. 0.5) then
             m_limits = (/25.5_8,27.5_8/)
         end if
         if (group .lt. 0.5) then
C            Classical belt object                 
             a_limits = (/40.0_8,50.0_8/)
             e_limits = (/0.0_8,0.5_8/)
             i_limits = (/0.0_8,Pi/6.0_8/)
         end if
         if (group .lt. 0.3) then
             m_limits = (/25.5_8,27.5_8/)
         end if

         niter = niter + 1
         a = a_limits(1) + ran3(seed)*(a_limits(2)-a_limits(1))
         e = e_limits(1) + ran3(seed)*(e_limits(2)-e_limits(1))
         inc = i_limits(1) + ran3(seed)*(i_limits(2)-i_limits(1))
         ran_mag = m_limits(1) + ran3(seed)*(m_limits(2)-m_limits(1))
         node = TwoPi*ran3(seed)
         peri = TwoPi*ran3(seed)
         M = TwoPi*ran3(seed)
C     compute a g-i color as randomly distributed between 0.5 and 1.5
C     color range in based of Oflek 2012 (SDDS colors of TNOS)
         gi = ran3(seed) + 0.5
            
            
         call pos_cart(a, e, inc, node, peri, M, pos(1),
     $        pos(2), pos(3))
         call RADECeclXV (pos, obs_pos, delta, ra, dec, ierr)
C     Given this random orbit determine if its on the image.

C     This loop selects sources that are within RADIUS for a field 
         obs_ra = ra_cen*drad
         obs_dec = dec_cen*drad
         if (radius(ra, dec, obs_ra, obs_dec) .lt. obs_radius) then
            nobjects = nobjects +1

C     Compute the Ground Based magnitude
            r = sqrt(pos(1)**2 + pos(2)**2 + pos(3)**2)
C     Set H based on randomize value of mag
            H = ran_mag -5*LOG10(r) - 5*LOG10(delta)

C     This orbit / object is inside our sample. 
C     Compute circumstances at current observation.

            fake_M = MA(M, epoch_M, a, fake_jday)
            call pos_cart(a, e, inc, node, peri, fake_M, pos(1),
     $           pos(2), pos(3))
            call RADECeclXV (pos, fake_obs_pos, fake_delta, 
     $           fake_ra, fake_dec, ierr)
            r = sqrt(pos(1)**2 + pos(2)**2 + pos(3)**2)
            call AppMag(r, delta, fake_ros, H, gb, alpha, 
     $           mag, ierr) 

C     Compute position 1 day later to get sky motion rate
            fake_M = MA(M, epoch_M, a, fake_jday+1.0)
            call pos_cart(a, e, inc, node, peri, fake_M, pos(1),
     $           pos(2), pos(3))

            call RADECeclXV (pos, fake_obs_pos2, fake_delta2, 
     $           fake_ra2, fake_dec2, ierr)

C     account for the fact we're doing this wrong. 
            dra = fake_ra2 - fake_ra
            if ( abs(dra) .gt. 1) then
               delta = TwoPi
               if ( dra .gt. 1 ) then
                  delta = -1*delta
               end if
               dra = dra + delta
            end if
            dra = dra*cos(fake_dec2)/24.0
            ddec = (fake_dec2-fake_dec)/24.0
            object_number = niter


            write (output_lun, 9001) object_number,
     $           fake_ra/drad, fake_dec/drad, fake_delta, mag,
     $           a, e, inc/drad, node/drad, peri/drad,
     $           fake_M/drad, H, 3600*dra/drad, 3600.0*ddec/drad,
     $           r, gi
            call flush(output_lun)
         end if
      end do

      STOP

 9001 format (i10, 15(f16.6,1x))
 9002 format (a10, 15(a16,1x))

      end program xx

      real*8 function radius(ra1, dec1, ra2, dec2)
      real*8 ra1, dec1, ra2, dec2
      
      radius = acos(sin(dec1)*sin(dec2) + 
     $     cos(dec1)*cos(dec2)*cos(ra1-ra2))
      return
      end function radius

      real*8 function MA(M, epoch_M, a, jday)
C     Move the Mean Annomally from jday to epoch_M

      real*8 M
      real*8 jday
      real*8 nu, TwoPi, epoch_M, a
      parameter (TwoPi=3.141592653589793238d0*2d0, nu=TwoPi/365.25d0)

      MA = M + (jday-epoch_M)*((nu/a**1.5))
      MA = MA - int(MA/TwoPi)*TwoPi
      return
      end function MA
      


      subroutine usage(message, ierr)
      character*80 message
      integer*4 ierr

      write(0,*) trim(message)
      write(0,*) ""

      write(0,*) "Usage: ClassyMakeFake FIELD RA DEC JD NITER"
      write(0,*) " FIELD -- Name of the field"
      write(0,*) " RA    -- RA of field center, in degrees"
      write(0,*) " DEC   -- RA of field center, in degrees"
      write(0,*) " JD    -- Julian Date assocaited to RA/DEC position"
      write(0,*) " JD    -- Julian Date for image to plant into"
      write(0,*) " NITER -- Number of KBOs to generate"
      write(0,*) ""

      call exit(ierr)
      end subroutine usage


      subroutine parse_args(name, ra, dec, obs_jd, fake_jd, max_objects)

C     Get the JD we will generate FAKE objects positions for.
C     Usage;  MakeFake obs_ra obs_dec obs_jday fake_jday number

      real*8 ra, dec, obs_jd, fake_jd
      character*80 message, command
      character*120 in_str, name
      integer*4 length
      integer*4 max_objects
      integer*4 nargs
      integer*4 argn


      nargs = COMMAND_ARGUMENT_COUNT()
      message = 'Wrong number of arguments'
      if (nargs .ne. 6) CALL usage(message, -1)

      argn = 0
      length = 80
      CALL GET_COMMAND_ARGUMENT(argn, command, length, ierr)
      if ( ierr .ne. 0 ) GOTO 1000

      argn = 1
      length = 120
      CALL GET_COMMAND_ARGUMENT(argn, name, length, ierr)
      if ( ierr .ne. 0 ) GOTO 1000

      argn = 2
      length = 80
      CALL GET_COMMAND_ARGUMENT(argn, in_str, length, ierr)
      if ( ierr .ne. 0 ) GOTO 1000
      read (in_str, *, err=1000) ra

      argn = 3
      CALL GET_COMMAND_ARGUMENT(argn, in_str, length, ierr)
      if (ierr .ne. 0) GOTO 1000
      read (in_str, *, err=1000) dec

      argn = 4
      CALL GET_COMMAND_ARGUMENT(argn, in_str, length, ierr)
      if (ierr .ne. 0) GOTO 1000
      read (in_str, *, err=1000) obs_jd

      argn = 5
      CALL GET_COMMAND_ARGUMENT(argn, in_str, length, ierr)
      if (ierr .ne. 0) GOTO 1000
      read (in_str, *, err=1000) fake_jd

C     Get the number of objects to generate
      argn = 6
      CALL GET_COMMAND_ARGUMENT(argn, in_str, length, ierr)
      if (ierr .ne. 0) GOTO 1000
      read (in_str, *, err=1000) max_objects

      write(0,*) "Executing command: ",command
      write(0,*) "Field Centre", ra, dec
      write(0,*) "Taken on JD:", obs_jd
      write(0,*) "Propogating elements to:", fake_jd
      write(0,*) "Number of objects:", max_objects
      RETURN

 1000 CONTINUE

      write(message, 1001) "Error parsing argument", argn
      message = trim(message)
      CALL usage(message, -1)
 1001 FORMAT('A20,I10')
      end subroutine parse_args

